{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37b00429-be42-44a1-883a-66945f73ae2f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# this tool takes in a FASTA file of dna and translates it directly into codons and their respective amino acids.\n",
    "# this block contains the main() function used as an entry point to the whole code and its multiple functions.\n",
    "def main():\n",
    "    filename = input(\"enter sequence filename: \")\n",
    "    sequence = read_file(filename)\n",
    "    codons = get_codons(sequence)\n",
    "    aa_list = trans_codons(codons)\n",
    "    \n",
    "    print(\"stripped sequence:\",sequence)\n",
    "    print(\"the codons are: \",codons)\n",
    "    print(\"the amino acids are: \",aa_list)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "017e076c-8e45-4965-bb20-c4c25b86b470",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function reads the file line by line, strips any space,comma or unwanted noise in the sequence and puts it into a string.\n",
    "def read_file(filename):\n",
    "    with open(filename,'r') as file:\n",
    "        lines = file.readlines()\n",
    "        \n",
    "        \n",
    "        sequence = \" \"\n",
    "        \n",
    "        for line in lines:\n",
    "            if not line.startswith(\">\"):\n",
    "                sequence += line.strip()\n",
    "        return sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "216604b9-671f-45b6-84c4-d00644f3b695",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this function then converts the sequence into triplet codons.\n",
    "def get_codons(sequence):\n",
    "    sequence = sequence.strip().upper()\n",
    "    \n",
    "    codons = []\n",
    "\n",
    "    remainder = len(sequence) % 3\n",
    "    if remainder != 0:\n",
    "        sequence = sequence[:-remainder]\n",
    "\n",
    "    for i in range(0,len(sequence),3):\n",
    "        codon = sequence[i:i+3]\n",
    "        if len(codon) == 3:\n",
    "            codons.append(codon)\n",
    "    return codons    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b75a5dd0-cfa4-4ac6-ba5f-2ca9f5dd83ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# here a dictionary is made with codons and their respectice amino acids so that the codons are translated to provide protein skeleton by the last function.\n",
    "genetic_code = {\n",
    "    \"TTT\": \"F\", \"TTC\": \"F\",\n",
    "    \"TTA\": \"L\", \"TTG\": \"L\",\n",
    "    \"CTT\": \"L\", \"CTC\": \"L\", \"CTA\": \"L\", \"CTG\": \"L\",\n",
    "\n",
    "    \"ATT\": \"I\", \"ATC\": \"I\", \"ATA\": \"I\",\n",
    "    \"ATG\": \"M\",\n",
    "\n",
    "    \"GTT\": \"V\", \"GTC\": \"V\", \"GTA\": \"V\", \"GTG\": \"V\",\n",
    "\n",
    "    \"TCT\": \"S\", \"TCC\": \"S\", \"TCA\": \"S\", \"TCG\": \"S\",\n",
    "    \"AGT\": \"S\", \"AGC\": \"S\",\n",
    "\n",
    "    \"CCT\": \"P\", \"CCC\": \"P\", \"CCA\": \"P\", \"CCG\": \"P\",\n",
    "\n",
    "    \"ACT\": \"T\", \"ACC\": \"T\", \"ACA\": \"T\", \"ACG\": \"T\",\n",
    "\n",
    "    \"GCT\": \"A\", \"GCC\": \"A\", \"GCA\": \"A\", \"GCG\": \"A\",\n",
    "\n",
    "    \"TAT\": \"Y\", \"TAC\": \"Y\",\n",
    "\n",
    "    \"AAA\": \"K\", \"AAG\": \"K\",\n",
    "    \n",
    "    \"CAT\": \"H\", \"CAC\": \"H\",\n",
    "\n",
    "    \"CAA\": \"Q\", \"CAG\": \"Q\",\n",
    "\n",
    "    \"AAT\": \"N\", \"AAC\": \"N\",\n",
    "\n",
    "\n",
    "    \"GAT\": \"D\", \"GAC\": \"D\",\n",
    "\n",
    "    \"GAA\": \"E\", \"GAG\": \"E\",\n",
    "\n",
    "    \"TGT\": \"C\", \"TGC\": \"C\",\n",
    "\n",
    "    \"TGG\": \"W\",\n",
    "\n",
    "    \"CGT\": \"R\", \"CGC\": \"R\", \"CGA\": \"R\", \"CGG\": \"R\",\n",
    "    \"AGA\": \"R\", \"AGG\": \"R\",\n",
    "\n",
    "    \"GGT\": \"G\", \"GGC\": \"G\", \"GGA\": \"G\", \"GGG\": \"G\",\n",
    "\n",
    "    \"TAA\": \"*\", \"TAG\": \"*\", \"TGA\": \"*\"\n",
    "}\n",
    "\n",
    "def trans_codons(codons):\n",
    "    \n",
    "    aa_list = []\n",
    "\n",
    "    for codon in codons:\n",
    "        amino_acids = genetic_code[codon]\n",
    "        aa_list.append(amino_acids)\n",
    "    return aa_list    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16d2c0fd-8673-465d-b299-193f3bcc18e3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8942f4f-3131-496b-b7d3-b47c67a3cf6d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c379d43-ebb8-436f-9bc4-f3994ba6fef4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
